### ISSUE TITLE
using ExponentialMovingAverage differs in CPU & GPU

### ISSUE THREAD

**Issue by DarcyCode** [tensorflow/tensorflow#13461]
I use BN in cnn,here is the code

```
def batch_norm(x, beta, gamma, train_phase, scope='bn', decay=0.9, eps=1e-5):

    with tf.variable_scope(scope):        
        batch_mean, batch_var = tf.nn.moments(x, [0, 1, 2], name='moments')
        ema = tf.train.ExponentialMovingAverage(decay=decay)

        def mean_var_with_update():
            ema_apply_op = ema.apply([batch_mean, batch_var])
            with tf.control_dependencies([ema_apply_op]):
                return tf.identity(batch_mean), tf.identity(batch_var)

        mean, var = tf.cond(train_phase, mean_var_with_update, lambda: (ema.average(batch_mean), ema.average(batch_var)))
        normed = tf.nn.batch_normalization(x, mean, var, beta, gamma, eps)
    return normed
```

when i print variables in CPU & GPU, it differs:
In CPU it shows:
`bn_1/bn_1/moments/Squeeze/ExponentialMovingAverage (DT_FLOAT) [32]
bn_1/bn_1/moments/Squeeze_1/ExponentialMovingAverage (DT_FLOAT) [32]
bn_2/bn_2/moments/Squeeze/ExponentialMovingAverage (DT_FLOAT) [64]
bn_2/bn_2/moments/Squeeze_1/ExponentialMovingAverage (DT_FLOAT) [64]
bn_3/bn_3/moments/Squeeze/ExponentialMovingAverage (DT_FLOAT) [64]
bn_3/bn_3/moments/Squeeze_1/ExponentialMovingAverage (DT_FLOAT) [64]
bn_4/bn_4/moments/Squeeze/ExponentialMovingAverage (DT_FLOAT) [64]
bn_4/bn_4/moments/Squeeze_1/ExponentialMovingAverage (DT_FLOAT) [64]`
while in GPU it shows:
`bn_1/bn_1/moments/moments_1/mean/ExponentialMovingAverage (DT_FLOAT) [32]
bn_1/bn_1/moments/moments_1/variance/ExponentialMovingAverage (DT_FLOAT) [32]
bn_2/bn_2/moments/moments_1/mean/ExponentialMovingAverage (DT_FLOAT) [64]
bn_2/bn_2/moments/moments_1/variance/ExponentialMovingAverage (DT_FLOAT) [64]
bn_3/bn_3/moments/moments_1/mean/ExponentialMovingAverage (DT_FLOAT) [64]
bn_3/bn_3/moments/moments_1/variance/ExponentialMovingAverage (DT_FLOAT) [64]
bn_4/bn_4/moments/moments_1/mean/ExponentialMovingAverage (DT_FLOAT) [64]
bn_4/bn_4/moments/moments_1/variance/ExponentialMovingAverage (DT_FLOAT) [64]`

therefore, when i load a CPU-trainned ckeckpoint in GPU, it turns out some errors like "variables not found in file" etc. Can someone solve this problem? thanks a lot.

**Comment by cy89** [comment]
@asimshankar it's not clear to me whether this behavior is working as intended (you can't read a CPU-trained checkpoint on a GPU), or whether this is supposed to work. Can you please comment?

**Comment by asimshankar** [comment]
The checkpoint should certainly be loadable. The provided snippet isn't sufficient to help reproduce the problem. @DarcyCode could you please fill in the complete issue template, including instructions to reproduce the problem?

Simply using the `batch_norm` function you've provided above and with the following snippet, I do not see any difference in variables names with or without GPU:

```python
import tensorflow as tf

def batch_norm(x, beta, gamma, train_phase, scope='bn', decay=0.9, eps=1e-5):
    with tf.variable_scope(scope):        
        batch_mean, batch_var = tf.nn.moments(x, [0, 1, 2], name='moments')
        ema = tf.train.ExponentialMovingAverage(decay=decay)

        def mean_var_with_update():
            ema_apply_op = ema.apply([batch_mean, batch_var])
            with tf.control_dependencies([ema_apply_op]):
                return tf.identity(batch_mean), tf.identity(batch_var)

        mean, var = tf.cond(train_phase, mean_var_with_update, lambda: (ema.average(batch_mean), ema.average(batch_var)))
        normed = tf.nn.batch_normalization(x, mean, var, beta, gamma, eps)
    return normed

def build_graph():
  x = tf.ones([1, 1, 1])
  train_phase = tf.constant(True)
  batch_norm(x, 1., 1., train_phase)
  print(tf.global_variables())

with tf.Graph().as_default():
  with tf.device('/device:CPU:0'):
    build_graph()

print('-----')

with tf.Graph().as_default():
  with tf.device('/device:GPU:0'):
    build_graph()
```

**Comment by DarcyCode** [comment]
@asimshankar Thanks for answering my questions. I define a CNN with BN on a computer(no-GPU) with CPU-version tensorflow r1.3. Code is below:
`X = tf.placeholder(tf.float32, [None, LENGTH],name='X')
Y = tf.placeholder(tf.float32, [None, MAX_CAPTCHA*CHAR_SET_LEN],name='Y')
keep_prob = tf.placeholder(tf.float32,name='keep_prob') # dropout
train_phase = tf.placeholder(tf.bool, name='train_phase')

def batch_norm(x, beta, gamma, train_phase, scope='bn', decay=0.9, eps=1e-5):
    with tf.variable_scope(scope):
        batch_mean, batch_var = tf.nn.moments(x, [0, 1, 2], name='moments')
        ema = tf.train.ExponentialMovingAverage(decay=decay)

        def mean_var_with_update():
            ema_apply_op = ema.apply([batch_mean, batch_var])
            with tf.control_dependencies([ema_apply_op]):
                return tf.identity(batch_mean), tf.identity(batch_var)

        mean, var = tf.cond(train_phase, mean_var_with_update, lambda: (ema.average(batch_mean), ema.average(batch_var)))
        normed = tf.nn.batch_normalization(x, mean, var, beta, gamma, eps)
    return normed

def crack_captcha_cnn(w_alpha=0.01, b_alpha=0.1):
    x = tf.reshape(X, shape=[-1, IMAGE_HEIGHT, IMAGE_WIDTH, 1])
    
    # 4 conv layer
    w_c1 = tf.Variable(w_alpha*tf.random_normal([5, 5, 1, 32]))
    b_c1 = tf.Variable(b_alpha*tf.random_normal([32]))
    conv1 = tf.nn.bias_add(tf.nn.conv2d(x, w_c1, strides=[1, 1, 1, 1], padding='SAME'), b_c1)
    conv1 = batch_norm(conv1, tf.constant(0.0, shape=[32]), tf.random_normal(shape=[32], mean=1.0, stddev=0.02), train_phase, scope='bn_1')
    conv1 = tf.nn.relu(conv1)
    conv1 = tf.nn.max_pool(conv1, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')
    conv1 = tf.nn.dropout(conv1, keep_prob)
 
    w_c2 = tf.Variable(w_alpha*tf.random_normal([5, 5, 32, 64]))
    b_c2 = tf.Variable(b_alpha*tf.random_normal([64]))
    conv2 = tf.nn.bias_add(tf.nn.conv2d(conv1, w_c2, strides=[1, 1, 1, 1], padding='SAME'), b_c2)
    conv2 = batch_norm(conv2, tf.constant(0.0, shape=[64]), tf.random_normal(shape=[64], mean=1.0, stddev=0.02), train_phase, scope='bn_2')
    conv2 = tf.nn.relu(conv2)
    conv2 = tf.nn.max_pool(conv2, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')
    conv2 = tf.nn.dropout(conv2, keep_prob)
 
    w_c3 = tf.Variable(w_alpha*tf.random_normal([3, 3, 64, 64]))
    b_c3 = tf.Variable(b_alpha*tf.random_normal([64]))
    conv3 = tf.nn.bias_add(tf.nn.conv2d(conv2, w_c3, strides=[1, 1, 1, 1], padding='SAME'), b_c3)
    conv3 = batch_norm(conv3, tf.constant(0.0, shape=[64]), tf.random_normal(shape=[64], mean=1.0, stddev=0.02), train_phase, scope='bn_3')
    conv3 = tf.nn.relu(conv3)
    conv3 = tf.nn.max_pool(conv3, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')
    conv3 = tf.nn.dropout(conv3, keep_prob)

    w_c4 = tf.Variable(w_alpha*tf.random_normal([3, 3, 64, 64]))
    b_c4 = tf.Variable(b_alpha*tf.random_normal([64]))
    conv4 = tf.nn.bias_add(tf.nn.conv2d(conv3, w_c4, strides=[1, 1, 1, 1], padding='SAME'), b_c4)
    conv4 = batch_norm(conv4, tf.constant(0.0, shape=[64]), tf.random_normal(shape=[64], mean=1.0, stddev=0.02), train_phase, scope='bn_4')
    conv4 = tf.nn.relu(conv4)
    conv4 = tf.nn.max_pool(conv4, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')
    conv4 = tf.nn.dropout(conv4, keep_prob)
     
    # Fully connected layer
    w_d = tf.Variable(w_alpha*tf.random_normal([int(LENGTH/4), 1024]))
    b_d = tf.Variable(b_alpha*tf.random_normal([1024]))
    dense = tf.reshape(conv4, [-1, w_d.get_shape().as_list()[0]])
    dense = tf.nn.relu(tf.add(tf.matmul(dense, w_d), b_d))
    dense = tf.nn.dropout(dense, keep_prob)
 
    w_out = tf.Variable(w_alpha*tf.random_normal([1024, MAX_CAPTCHA*CHAR_SET_LEN]))
    b_out = tf.Variable(b_alpha*tf.random_normal([MAX_CAPTCHA*CHAR_SET_LEN]))
    out = tf.add(tf.matmul(dense, w_out), b_out, name='out')
    return out

def train_crack_captcha_cnn():
    output = crack_captcha_cnn()

    loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits=output, labels=Y))
    optimizer = tf.train.AdamOptimizer(learning_rate=0.001).minimize(loss)
 
    predict = tf.reshape(output, [-1, MAX_CAPTCHA, CHAR_SET_LEN],name='predict')
    max_idx_p = tf.argmax(predict, 2)
    max_idx_l = tf.argmax(tf.reshape(Y, [-1, MAX_CAPTCHA, CHAR_SET_LEN]), 2)
    correct_pred = tf.equal(max_idx_p, max_idx_l)
    accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32), name='accuracy')
 
    saver = tf.train.Saver()
    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        for i in tf.global_variables():
            print(i)
        return`
When print tf.global_variables(), it prints
`<tf.Variable 'Variable:0' shape=(5, 5, 1, 32) dtype=float32_ref>
<tf.Variable 'Variable_1:0' shape=(32,) dtype=float32_ref>
<tf.Variable 'bn_1/bn_1/moments/Squeeze/ExponentialMovingAverage:0' shape=(32,) dtype=float32_ref>
<tf.Variable 'bn_1/bn_1/moments/Squeeze_1/ExponentialMovingAverage:0' shape=(32,) dtype=float32_ref>
<tf.Variable 'Variable_2:0' shape=(5, 5, 32, 64) dtype=float32_ref>
<tf.Variable 'Variable_3:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_2/bn_2/moments/Squeeze/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_2/bn_2/moments/Squeeze_1/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'Variable_4:0' shape=(3, 3, 64, 64) dtype=float32_ref>
<tf.Variable 'Variable_5:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_3/bn_3/moments/Squeeze/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_3/bn_3/moments/Squeeze_1/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'Variable_6:0' shape=(3, 3, 64, 64) dtype=float32_ref>
<tf.Variable 'Variable_7:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_4/bn_4/moments/Squeeze/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_4/bn_4/moments/Squeeze_1/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'Variable_8:0' shape=(3200, 1024) dtype=float32_ref>
<tf.Variable 'Variable_9:0' shape=(1024,) dtype=float32_ref>
<tf.Variable 'Variable_10:0' shape=(1024, 200) dtype=float32_ref>
<tf.Variable 'Variable_11:0' shape=(200,) dtype=float32_ref>
<tf.Variable 'beta1_power:0' shape=() dtype=float32_ref>
<tf.Variable 'beta2_power:0' shape=() dtype=float32_ref`
However, when i define the same net on a computer(a GPU) with GPU-version tensorflow r1.2, it prints:
`<tf.Variable 'Variable:0' shape=(5, 5, 1, 32) dtype=float32_ref>
<tf.Variable 'Variable_1:0' shape=(32,) dtype=float32_ref>
<tf.Variable 'bn_1/bn_1/moments/moments_1/mean/ExponentialMovingAverage:0' shape=(32,) dtype=float32_ref>
<tf.Variable 'bn_1/bn_1/moments/moments_1/variance/ExponentialMovingAverage:0' shape=(32,) dtype=float32_ref>
<tf.Variable 'Variable_2:0' shape=(5, 5, 32, 64) dtype=float32_ref>
<tf.Variable 'Variable_3:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_2/bn_2/moments/moments_1/mean/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_2/bn_2/moments/moments_1/variance/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'Variable_4:0' shape=(3, 3, 64, 64) dtype=float32_ref>
<tf.Variable 'Variable_5:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_3/bn_3/moments/moments_1/mean/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_3/bn_3/moments/moments_1/variance/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'Variable_6:0' shape=(3, 3, 64, 64) dtype=float32_ref>
<tf.Variable 'Variable_7:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_4/bn_4/moments/moments_1/mean/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'bn_4/bn_4/moments/moments_1/variance/ExponentialMovingAverage:0' shape=(64,) dtype=float32_ref>
<tf.Variable 'Variable_8:0' shape=(3200, 1024) dtype=float32_ref>
<tf.Variable 'Variable_9:0' shape=(1024,) dtype=float32_ref>
<tf.Variable 'Variable_10:0' shape=(1024, 200) dtype=float32_ref>
<tf.Variable 'Variable_11:0' shape=(200,) dtype=float32_ref>`

**Comment by asimshankar** [comment]
@DarcyCode : The snippet you provided above doesn't seem to be self-contained, so I'm not sure if there are other things at play here.

Could you provide a small, self-contained snippet that reproduces the problem?

**Comment by DarcyCode** [comment]
All code is pasted above. I think the problem lies in function batch_norm(). It is saved differently in different machine. Or my computer is too old for tensorflow....

**Comment by angersson** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
